crates/aptos-dkg/src/range_proofs/dekart_univariate_v2.rs

```diff
@@ -0,0 +1,812 @@
+            ell + 2,
+        );
+
+        let mu = mus.pop().expect("mus must have at least one element");
+        let mu_h = mus.pop().expect("mus must have at least two elements");
+
+        (mu, mu_h, mus)
+    }
+
+    #[allow(non_snake_case)]
+    pub(crate) fn get_gamma_challenge<E: Pairing>(
+        fs_transcript: &mut Transcript,
+        roots_of_unity: &Vec<E::ScalarField>,
+    ) -> E::ScalarField {
+        loop {
+            let gamma =
+                <Transcript as RangeProof<E, Proof<E>>>::challenge_from_verifier(fs_transcript);
+            if !roots_of_unity.contains(&gamma) {
+                return gamma;
+            }
+        }
+    }
+}
+
+pub mod two_term_msm {
+    // TODO: maybe fixed_base_msms should become a folder, put its code inside mod.rs? Then put this inside of that folder?
+    use super::*;
+    use crate::sigma_protocol::homomorphism::fixed_base_msms;
+    use aptos_crypto_derive::SigmaProtocolWitness;
+    pub use sigma_protocol::homomorphism::TrivialShape as CodomainShape;
+
+    #[derive(CanonicalSerialize, Clone, Debug, PartialEq, Eq)]
+    pub struct Homomorphism<E: Pairing> {
+        // This is rather similar to a Pedersen commitment
+        pub base_1: E::G1Affine,
+        pub base_2: E::G1Affine,
+    }
+
+    #[derive(SigmaProtocolWitness, CanonicalSerialize, CanonicalDeserialize, Clone)]
+    pub struct Witness<E: Pairing> {
+        pub kzg_randomness: Scalar<E>,
+        pub hiding_kzg_randomness: Scalar<E>,
+    }
+
+    impl<E: Pairing> homomorphism::Trait for Homomorphism<E> {
+        type Codomain = CodomainShape<E::G1>;
+        type Domain = Witness<E>;
+
+        fn apply(&self, input: &Self::Domain) -> Self::Codomain {
+            // Not doing `self.apply_msm(self.msm_terms(input))` because E::G1::msm is slower! `msm_terms()` is still useful for verification though
+            CodomainShape(
+                self.base_1 * input.kzg_randomness.0 + self.base_2 * input.hiding_kzg_randomness.0,
+            )
+        }
+    }
+
+    impl<E: Pairing> fixed_base_msms::Trait for Homomorphism<E> {
+        type Base = E::G1Affine;
+        type CodomainShape<T>
+            = CodomainShape<T>
+        where
+            T: CanonicalSerialize + CanonicalDeserialize + Clone;
+        type MsmInput = fixed_base_msms::MsmInput<Self::Base, Self::Scalar>;
+        type MsmOutput = E::G1;
+        type Scalar = E::ScalarField;
+
+        fn msm_terms(&self, input: &Self::Domain) -> Self::CodomainShape<Self::MsmInput> {
+            let mut scalars = Vec::with_capacity(2);
+            scalars.push(input.kzg_randomness.0);
+            scalars.push(input.hiding_kzg_randomness.0);
+
+            let mut bases = Vec::with_capacity(2);
+            bases.push(self.base_1);
+            bases.push(self.base_2);
+
+            CodomainShape(fixed_base_msms::MsmInput { bases, scalars })
+        }
+
+        fn msm_eval(bases: &[Self::Base], scalars: &[Self::Scalar]) -> Self::MsmOutput {
```

Comment:
So verification uses this, which is a size-2 G1 MSM, right?

But before you said that this will be slower than naive? If so, why do it here?